ZenPacks/zenoss/LinuxMonitor/parsers/linux/diskstats.py

```diff
@@ -1,140 +1,196 @@
-                    log.warning("%s ignored, invalid regex specified: %s", self.HARD_DISK_MATCH_PROPERTY, e)
-
-            # parse command output to extract values
-            log.debug("about to parse: %s", cmd.result.output)
-            valueMap = _parseOutput(cmd.result.output, hdFilterFn)
-
-            # save (datapoint, value) tuples to result values
-            result.values.extend((pointsMap[key], valueMap[key]) for key in destkeys)
-            log.debug("parsed values: %s", zip(destkeys, (x[1] for x in result.values[-len(destkeys):])))
+        datapointMap = dict((dp.id, dp) for dp in cmd.points)
+        if any((x in datapointMap for x in [SS_IO_RAW_RECEIVED, SS_IO_RAW_SENT])):
+            parser = StorageDeviceParser(datapointMap)
+            scanner = STORAGE_DEVICE
+        else:
+            parser = PartitionParser(cmd.points)
+            scanner = PARTITION
+
+        LOG.debug("Parser class: %s, Scanner: %s", parser.__class__, scanner)
+
+        parts = cmd.result.output.split(self.splitter)
+        for part in parts:
+            match = re.search(scanner, part)
+            if match:
+                parser.parse(match, part)
+
+        metrics = parser.getMetrics()
+        LOG.debug("Parsed metrics: %s", metrics)
+        if metrics:
+            result.values.extend(metrics)
 
         return result
+
+
+class Parser(object):
+
+    scanners = []
+
+    def __init__(self):
+        self._metrics = []
+
+    def parse(self, datapointMap, part):
+        for search in self.scanners:
+            match = re.search(search, part)
+            if match:
+                for name, value in match.groupdict().items():
+                    dp = datapointMap.get(name, None)
+                    if dp:
+                        if value in ('-', ''):
+                            value = 0
+                        self._metrics.append((dp, float(value)))
+
+    def getMetrics(self):
+        return self._metrics
+
+
+class PartitionParser(Parser):
+
+    scanners = [
+        r''.join([PARTITION, TOTAL_READS, NUM_READS_MERGED, TOTAL_SECTORS_READ,
+                  TOTAL_MS_READING, TOTAL_WRITES, NUM_WRITES_MERGED,
+                  TOTAL_SECTORS_WRITTEN, TOTAL_MS_WRITING, NUM_IO_IN_PROGESS,
+                  TOTAL_MS_IO, WEIGHTED_MS_PER_IO]),
+    ]
+
+    def __init__(self, datapoints):
+        super(PartitionParser, self).__init__()
+        self._components = {}
+        for dp in datapoints:
+            dp.component = dp.data['componentScanValue']
+            points = self._components.setdefault(dp.component, {})
+            points[dp.id] = dp
+
+    def parse(self, match, part):
+        component = match.groupdict()['component'].strip()
+        datapointMap = self._getDatapointMap(component)
+        if datapointMap:
+            super(PartitionParser, self).parse(datapointMap, part)
+
+    def _getDatapointMap(self, component):
+        for k, v in self._components.iteritems():
+            if component in k:
```

Comment:
A substring match here is too optimistic. A line for "xvdb" could return the datapoint map for "dev_xvdb1". The match between lines in /proc/diskstats and components needs to be specific.